src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,192 @@
+        /// </summary>
+        public bool ShouldCommitCompletion(
+            IAsyncCompletionSession session,
+            SnapshotPoint location,
+            char typedChar,
+            CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+            if (session.Properties.TryGetProperty<ImmutableArray<char>>(CompletionSource.ExcludedCommitCharacters, out var excludedCommitCharacter))
+            {
+                if (excludedCommitCharacter.Contains(typedChar))
+                {
+                    return false;
+                }
+            }
+
+            return PotentialCommitCharacters.Contains(typedChar);
+        }
+
+        public AsyncCompletionData.CommitResult TryCommit(
+            IAsyncCompletionSession session,
+            ITextBuffer subjectBuffer,
+            VSCompletionItem item,
+            char typeChar,
+            CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+            if (completionService == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(CompletionSource.RoslynItem, out var roslynItem))
+            {
+                // This isn't an item we provided (e.g. Razor). Let the editor handle it normally.
+                return CommitResultUnhandled;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(session.ApplicableToSpan.TextBuffer.CurrentSnapshot) + typeChar;
+            if (Controller.IsFilterCharacter(roslynItem, typeChar, filterText))
+            { 
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            var serviceRules = completionService.GetRules();
+
+            // We can be called before for ShouldCommitCompletion. However, that call does not provide rules applied for the completion item.
+            // Now we check for the commit charcter in the context of Rules that could change the list of commit characters.
+
+            // Tab, Enter and Null (call invoke commit) are always a commit character. 
+            if (typeChar != '\t' && typeChar != '\n' && typeChar != '\0' && !Controller.IsCommitCharacter(serviceRules, roslynItem, typeChar, filterText))
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            if (!item.Properties.TryGetProperty<ITextSnapshot>(CompletionSource.TriggerSnapshot, out var triggerSnapshot))
```

Comment:
1. definitely critical to have this.  however, should be doc'ed.  
2. you use generics here for all these cases, along with `out var`.  It might be cleaner to just do `out ITextSnapshot triggerSnapshot`.